### ISSUE TITLE
Kronos-base模型预测结果中volume取值含有负数

### ISSUE THREAD

**Issue by anbyaa** [shiyu-coder/Kronos#130]
```
import os
import pandas as pd
import torch
from tqdm import tqdm
import multiprocessing

from Kronos.model import Kronos, KronosTokenizer, KronosPredictor

def set_seed(seed):
    import random
    import numpy as np
    import torch

    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(seed)
set_seed(0)

lookback=512
time_freq = 1
file_dir = f"data/UsStock{time_freq}m"
available_dates = pd.read_csv("data/S&P500 Stock.csv", header=[0,1], index_col=0, parse_dates=True)
available_dates = list(available_dates[available_dates.index > pd.to_datetime("2020-01-01")].index)

@torch.no_grad
def oper(args):
    idx, file = args
    tokenizer = KronosTokenizer.from_pretrained("Kronos/NeoQuasar/Kronos-Tokenizer-base")
    model = Kronos.from_pretrained("Kronos/NeoQuasar/Kronos-base")

    predictor = KronosPredictor(model, tokenizer, device=f"cuda:{idx}")
    df = pd.read_csv(file_dir + "//" + file,index_col=0, header=0, parse_dates=True)
    output = pd.DataFrame([], index=[], columns = [])
    for i, date in enumerate(available_dates):
        pred_start = date + pd.Timedelta("9:30:00")
        pred_end = date + pd.Timedelta("10:00:00")
        y_timestamp = pd.date_range(start=pred_start, end=pred_end, freq=f'{time_freq}min')
        x_timestamp = df[df.index < pred_start].index[-lookback:]
        x_df = df.loc[x_timestamp]
        pred_df = predictor.predict(
            df=x_df,
            x_timestamp=x_timestamp,
            y_timestamp=y_timestamp,
            pred_len=len(y_timestamp),
            T=0.6,
            top_p=0.9,
            sample_count=10,
            verbose=True
        )
        output = pd.concat([output, pred_df])
        print(output.iloc[:, -10:])
        if i==3:
            break

if __name__ == "__main__":
    multiprocessing.set_start_method('spawn', force=True)
    with multiprocessing.Pool(1) as pool:
        outputs = pool.imap_unordered(oper, [(i%8, j) for (i, j) in enumerate(os.listdir(file_dir)) if j=="AAPL.csv"])
        with tqdm(total=len(os.listdir(file_dir))) as pbar:
            for output in outputs:
                pbar.update()

```
以上是我的调试代码

对于Kronos模型代码，我将Kronos\model\kronos.py中的predict方法修改为回传全部10次预测的结果，并将正则化方式修改为

`x_min, x_max = np.min(x, axis=0), np.max(x, axis=0)
x = (x - x_mean) / (x_max - x_min)
...
preds = preds * (x_max - x_min) + x_mean`

对于模型代码没有做其它重大修改。然后在模型输出结果中，volume取值居然出现了负数，数值如下：
                           volume                              ...                                          
                                0             1             2  ...             7             8             9
2020-01-02 09:30:00  31320.001953  31320.001953  31320.001953  ...  31320.001953  31320.001953  31320.001953
2020-01-02 09:31:00  29214.054688  29214.054688  29214.054688  ...  29214.054688  29214.054688  29214.054688
2020-01-02 09:32:00  27728.085938  27728.085938  27728.085938  ...  27728.085938  27728.085938  27728.085938
2020-01-02 09:33:00  26652.433594  26652.433594  26652.433594  ...  26652.433594  26652.433594  26652.433594
2020-01-02 09:34:00  25892.529297  25892.529297  25892.529297  ...  25892.529297  25892.529297  25892.529297
...                           ...           ...           ...  ...           ...           ...           ...
2020-01-06 09:56:00 -12893.912109 -16550.912109 -16359.072266  ... -16319.962891 -16330.998047 -16352.232422
2020-01-06 09:57:00 -15848.287109 -16380.607422 -16125.333984  ... -16123.029297 -16185.173828 -16170.072266
2020-01-06 09:58:00 -12825.810547 -16254.900391 -16002.384766  ... -16080.228516 -16076.431641 -16046.951172
2020-01-06 09:59:00 -14561.283203 -16130.787109 -15976.658203  ... -16039.353516 -15891.607422 -15991.556641
2020-01-06 10:00:00 -14611.287109 -15955.564453 -15891.591797  ... -15891.087891 -15737.966797 -15877.294922

不修改正则化方式（维持(x - x_mean) / (x_std)）时，数值如下：
                          volume                                         ...                                                    
                               0            1            2            3  ...            6            7            8            9
2020-01-02 09:30:00  -582.933594 -3916.482422  1397.369141  1397.369141  ...  1397.369141 -3916.482422  2101.509766  -582.933594
2020-01-02 09:31:00  -891.982422   451.603516  1473.970703  2136.818359  ...  1473.970703  1223.511719  1844.068359  2026.548828
2020-01-02 09:32:00 -1289.093750  1630.527344  2300.501953 -3404.095703  ...  1597.001953  1320.298828  1873.855469  1567.615234
2020-01-02 09:33:00 -1720.117188  1569.654297 -3192.085938  1304.884766  ...  1602.265625  1434.767578  1934.824219  1524.433594
2020-01-02 09:34:00 -2146.355469  1541.710938   408.351562  1390.353516  ...  1521.341797  1491.201172  1934.505859  1550.404297
...                          ...          ...          ...          ...  ...          ...          ...          ...          ...
2020-01-06 09:56:00 -5263.685547 -5224.826172 -5204.130859 -4864.599609  ... -4983.943359 -5270.318359 -5289.560547 -4857.474609
2020-01-06 09:57:00 -5393.966797 -5330.314453 -5307.611328 -4999.755859  ... -5131.091797 -5393.263672 -5456.341797 -4982.994141
2020-01-06 09:58:00 -5462.306641 -5381.751953 -5369.373047 -5066.736328  ... -5229.943359 -5461.396484 -5559.943359 -5047.669922
2020-01-06 09:59:00 -5542.654297 -5427.853516 -5423.919922 -5154.408203  ... -5328.947266 -5505.361328 -5630.072266 -5115.580078
2020-01-06 10:00:00 -5652.927734 -5506.966797 -5514.728516 -5263.626953  ... -5431.154297 -5580.498047 -5708.748047 -5225.966797

负数的比例有随机性。若将模型运行修改为运行在cpu上，负数比例有所下降，但也并没有消失。
请问你们在其它数据集上是否有观察到类似的情况，这一负数输出的意义是什么？如果模型无法直接将浮点数输出约束在正数范围，你们在训练时是否应该将输出值以np.exp()的方式进行处理，以确保量价数据的输出一定维持在正数范围呢？

**Comment by aaanby** [comment]
运行环境：Ubuntu 20.04.6 LTS python3.12.11
torch                                             2.8.0
pandas                                            2.2.2
multitasking                                      0.0.12
einops                                            0.8.1
safetensors                                       0.6.2

**Comment by anbyaa** [comment]
运行环境：Ubuntu 20.04.6 LTS python3.12.11
NVIDIA-SMI 575.57.08              Driver Version: 575.57.08      CUDA Version: 12.9 RTX3080 Ti
torch 2.8.0
pandas 2.2.2
multitasking 0.0.12
einops 0.8.1
safetensors 0.6.2

**Comment by shiyu-coder** [comment]
由于模型使用的是RevIN标准化的方式，volume确实可能存在负数，但是应该是少见情况，大部分情况下都是正常的。这主要是历史序列估计的mean和std并不是非常准确导致的。采用exp进行矫正理论上是可以的，但是实际操作可能又会出现极端值等情况，总的来说如何解决这件事情需要权衡和思考。

**Comment by anbyaa** [comment]
是的 当前我对结果直接采用abs处理以确保其范围正确，但在之后的改进中最好能采用其它方式进行约束和校正，以避免这一问题持续出现。
